### PR TITLE
[_shard] make ShardedTensor be a Tensor and nn.Parameter

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79825

Differential Revision: [D37707371](https://our.internmc.facebook.com/intern/diff/D37707371)

### CODE REVIEW

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 1]
```diff
@@ -67,7 +65,7 @@ def _register_remote_shards(sharded_tensor_id: int, rrefs: List[rpc.RRef[Shard]]
         else:
             sharded_tensor._register_remote_shards(rrefs, rpc_rank)
 
-class ShardedTensor(object):
+class ShardedTensor(torch.Tensor):
     """
     ShardedTensor is an abstraction to represent Tensors that are sharded
```
Comment (kumpera) [inline torch/distributed/_shard/sharded_tensor/api.py:70]: We should update the docstring to say it's a subclass of Tensor

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 2]
```diff
@@ -446,18 +519,24 @@ def _init_from_local_shards(
...
         global_sharded_tensor_metadata = build_global_metadata(gathered_metadatas)
+        tensor_properties = global_sharded_tensor_metadata.tensor_properties
 
         # STEP 3: Validation done, create the actual ShardedTensor and populate fields
         # prepare initialization
-        sharded_tensor = cls.__new__(cls)
+        spec = shard_spec._infer_sharding_spec_from_shards_metadata(
+            global_sharded_tensor_metadata.shards_metadata
+        )
+        sharded_tensor = cls.__new__(cls,
```
Comment (kumpera) [inline torch/distributed/_shard/sharded_tensor/api.py:529]: Can't we use `cls(...)` instead of `cls.__new__(...)`?

**Review by kumpera — APPROVED** [review]
LGTM but please check the multigpu jobs results before merging as they will have the ST test results in them.

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 3]
```diff
@@ -840,186 +928,12 @@ def local_shards(self) -> List[Shard]:
         """
         return self._local_shards
 
-    def size(self, dim: int = None) -> Union[torch.Size, int]:
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_shard/sharded_tensor/api.py:843]: Yes that's correct!

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37707371).

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37707371).

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_shard/sharded_tensor/api.py:612]: I didn't check if using `cls(...)` will trigger `__init__` or not, ideally we don't want to trigger `__init__` here and that's why I specified explicitly, I will check this as a follow up and remove `__new__` if `cls(...)` directly works for us :)

**Comment by janeyx99** [comment]
@pytorchbot revert -m "Sorry, reverting for breaking multigpu tests https://hud.pytorch.org/pytorch/pytorch/commit/9c32439a770845793107a584f17f4da70ccb5df0" -c nosignal

Please add ciflow/periodic label to your reland PR to trigger the multigpu tests.

**Pull request closed without merging**